certbot/tests/compat/filesystem_test.py

```diff
@@ -6,41 +6,162 @@
 
 
-class FilesystemTest(TempDirTestCase):
-    """Unit tests for filesystem module."""
-    @unittest.skipIf(os.name != 'nt', reason='Test specific to Windows security')
+@unittest.skipIf(os.name != 'nt', reason='Test specific to Windows security')
+class WindowsChmodTests(TempDirTestCase):
+    """Unit tests for Windows chmod function in filesystem module"""
+    def test_symlink_resolution(self):
+        probe_path = _create_probe(self.tempdir)  # This is 0o744 by default
+
+        link_path = os.path.join(self.tempdir, 'link')
+        os.symlink(probe_path, link_path)
+
+        ref_dacl_probe = _get_security_dacl(probe_path).GetSecurityDescriptorDacl()
+        ref_dacl_link = _get_security_dacl(link_path).GetSecurityDescriptorDacl()
+
+        # Removing the rights for `all`, we have at least one ACL less than in the case of 0o744.
+        filesystem.chmod(link_path, 0o700)
+
+        # Assert the real file is impacted, not the link.
+        cur_dacl_probe = _get_security_dacl(probe_path).GetSecurityDescriptorDacl()
+        cur_dacl_link = _get_security_dacl(link_path).GetSecurityDescriptorDacl()
+        self.assertFalse(filesystem._compare_dacls(ref_dacl_probe, cur_dacl_probe))  # pylint: disable=protected-access
+        self.assertTrue(filesystem._compare_dacls(ref_dacl_link, cur_dacl_link))  # pylint: disable=protected-access
+
+    def test_world_permission(self):
+        import win32security  # pylint: disable=import-error
+        probe_path = _create_probe(self.tempdir)
+
+        everybody = win32security.ConvertStringSidToSid('S-1-1-0')
+
+        filesystem.chmod(probe_path, 0o700)
+        dacl = _get_security_dacl(probe_path).GetSecurityDescriptorDacl()
+
+        self.assertFalse([dacl.GetAce(index) for index in range(0, dacl.GetAceCount())
+                          if dacl.GetAce(index)[2] == everybody])
+
+        filesystem.chmod(probe_path, 0o704)
+        dacl = _get_security_dacl(probe_path).GetSecurityDescriptorDacl()
+
+        self.assertTrue([dacl.GetAce(index) for index in range(0, dacl.GetAceCount())
+                         if dacl.GetAce(index)[2] == everybody])
+
+    def test_group_permissions_noop(self):
+        probe_path = _create_probe(self.tempdir)
+
+        filesystem.chmod(probe_path, 0o700)
+        ref_dacl_probe = _get_security_dacl(probe_path).GetSecurityDescriptorDacl()
+
+        filesystem.chmod(probe_path, 0o740)
+        cur_dacl_probe = _get_security_dacl(probe_path).GetSecurityDescriptorDacl()
+
+        self.assertTrue(filesystem._compare_dacls(ref_dacl_probe, cur_dacl_probe))  # pylint: disable=protected-access
+
+    def test_admin_permissions(self):
+        import win32security  # pylint: disable=import-error
+        probe_path = _create_probe(self.tempdir)
+
+        system = win32security.ConvertStringSidToSid('S-1-5-18')
+        admins = win32security.ConvertStringSidToSid('S-1-5-32-544')
+
+        filesystem.chmod(probe_path, 0o700)
```

Comment:
nit: Can we expand this test slightly by giving this file `0o400` permissions and then testing that not only do system and admins have ACEs, they have full permissions on the file?

Reply:
Done.